### PR TITLE
Viacheslav chernyshov

### PR DESCRIPTION
**Pull request by masteriza**
Dima, please check and confirm pull request. (=

### CODE REVIEW

File: web/src/main/java/ua/com/vertex/beans/Course.java [hunk 1]
```diff
@@ -1,66 +1,66 @@
...
+    private final String COURSE_TEACHER_NAME_MSG = "Teacher's name must be longer than 2 and less than 256 characters";
+    private final String SCHEDULE_MSG = "The schedule must be longer than 2 and less than 256 characters";
+    private final String NOTES_MSG = "The schedule must be up to 256 characters long";
 
     private int id;
-    private BigDecimal price;
-    private String name, teacherName, schedule, notes;
-    private LocalDateTime start;
+
+    @Size(min = 5, max = 256, message = COURSE_NAME_MSG)
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/beans/Course.java]: decrease min to 2 please

File: web/src/main/java/ua/com/vertex/controllers/AddCourseController.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+
+    @Autowired
+    public AddCourseController(CourseLogic courseLogic) {
+        this.courseLogic = courseLogic;
+    }
+
+    @GetMapping
+    public ModelAndView viewAddCourseForm() {
+        LOGGER.debug("First request to " + PAGE_JSP);
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/controllers/AddCourseController.java]: And what about second one?)

File: web/src/main/java/ua/com/vertex/controllers/AddCourseController.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                                          Course course, BindingResult bindingResult, ModelAndView modelAndView) {
+
+        if (bindingResult.hasErrors()) {
+            LOGGER.debug("The user entered incorrect course data -" + course);
+            modelAndView.setViewName(PAGE_JSP);
+        } else {
+            try {
+                course.setId(courseLogic.addCourse(course));
+                LOGGER.debug("Added a new course into DB" + course);
+                modelAndView.setViewName("redirect:" + ALL_COURSE_PAGE_JSP);
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/controllers/AddCourseController.java]: can you please include  redirect to the constant?

File: web/src/main/java/ua/com/vertex/controllers/ViewAllCoursesController.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import ua.com.vertex.utils.DataNavigator;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Controller
+@SessionAttributes(value = "viewAllCourses")
+public class ViewAllCoursesController {
+    private static final String ERROR_JSP = "error";
+    private static final String PAGE_JSP = "viewAllCourses";
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/controllers/ViewAllCoursesController.java]: or use this constant

File: web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java [hunk 5]
```diff
@@ -26,11 +34,92 @@
...
+
+        String query = "SELECT c.id, c.name, c.start, c.finished, c.price, c.teacher_name, c.schedule, c.notes " +
+                "FROM Courses c LIMIT :from, :offset";
+
+        MapSqlParameterSource parameters = new MapSqlParameterSource();
+        parameters.addValue("from", (dataNavigator.getCurrentNumberPage() - 1) * dataNavigator.getCurrentRowPerPage());
+        parameters.addValue("offset", dataNavigator.getCurrentRowPerPage());
+
+        try {
+            courses = jdbcTemplate.query(query, parameters, new CourseDaoImpl.CoursesRowMapping());
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java]: why don't you use lambda here?

File: web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java [hunk 6]
```diff
@@ -26,11 +34,92 @@
...
+        } catch (EmptyResultDataAccessException e) {
+            LOGGER.warn("Something went wrong", e);
+        }
+        return courses;
+    }
+
+    private static final class CoursesRowMapping implements RowMapper<Course> {
+        public Course mapRow(ResultSet resultSet, int i) throws SQLException {
+            Course course = new Course();
+            course.setId(resultSet.getInt("id"));
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java]: builder?

File: web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java [hunk 7]
```diff
@@ -26,11 +34,92 @@
...
+            course.setTeacherName(resultSet.getString("teacher_name"));
+            course.setSchedule(resultSet.getString("schedule"));
+            course.setNotes(resultSet.getString("notes"));
+            return course;
+        }
+    }
+
+    @Override
+    public int getQuantityCourses() throws SQLException {
+        LOGGER.debug("Get all courses list");
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java]: you write that you are going to get a list, but getting a quantity. It makes me sad =(

File: web/src/main/java/ua/com/vertex/logic/CourseLogicImpl.java [hunk 8]
```diff
@@ -7,23 +7,56 @@
...
+            dataNavigator.setNextPage(1);
+        } else {
+            dataNavigator.setCurrentNumberPage(dataNavigator.getNextPage());
+        }
+
+        try {
+            courses = courseDao.getAllCourses(dataNavigator);
+            int dataSize = courseDao.getQuantityCourses();
+
+            dataNavigator.setDataSize(dataSize);
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/logic/CourseLogicImpl.java]: some strange logic here that also changes incoming parameter. This approach is called "side effect". I do remember time when I was able to kill for that...

File: web/src/main/java/ua/com/vertex/logic/UserLogicImpl.java [hunk 9]
```diff
@@ -52,8 +52,29 @@ public void saveImage(int userId, byte[] image, String imageType) throws Excepti
     }
 
     @Override
-    public List<User> getAllUsers() throws SQLException {
-        return userDao.getAllUsers();
+    public List<User> getUsersPerPages(DataNavigator dataNavigator) {
+        List<User> users = null;
+        if (dataNavigator.getCurrentNumberPage() == dataNavigator.getNextPage()) {
+            dataNavigator.setCurrentNumberPage(1);
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/logic/UserLogicImpl.java]: same here

File: web/src/main/java/ua/com/vertex/utils/Aes.java [hunk 10]
```diff
@@ -19,23 +19,23 @@ private static String pad2(String n) {
     }
 
     private static String hex(byte[] bytes) {
-        String r = "";
+        StringBuilder r = new StringBuilder();
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/utils/Aes.java]: can you rename it please?

File: web/src/main/webapp/WEB-INF/views/viewAllCourses.jsp [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+        <div align="center">
+            <br/>
+            <br/>
+            <br/>
+            <a href="<c:url value="/addCourse"/>">Add course</a>
+            <br/>
+            <br/>
+            <br/>
+            <c:if test="${dataNavigator.currentNumberPage!=1}">
+                <a id="1" class="page" style="cursor: pointer;"><<</a>
```
Comment (backumich) [inline web/src/main/webapp/WEB-INF/views/viewAllCourses.jsp]: something strange here

**Review by backumich — CHANGES_REQUESTED** [review]

File: web/src/main/java/ua/com/vertex/beans/Course.java (hunk 1 above)
Comment (masteriza) [reply web/src/main/java/ua/com/vertex/beans/Course.java]: +

File: web/src/main/java/ua/com/vertex/controllers/AddCourseController.java (hunk 3 above)
Comment (masteriza) [reply web/src/main/java/ua/com/vertex/controllers/AddCourseController.java]: done

File: web/src/main/java/ua/com/vertex/controllers/AddCourseController.java (hunk 2 above)
Comment (masteriza) [reply web/src/main/java/ua/com/vertex/controllers/AddCourseController.java]: done

File: web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java (hunk 5 above)
Comment (masteriza) [reply web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java]: done

File: web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java (hunk 6 above)
Comment (masteriza) [reply web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java]: done

File: web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java (hunk 7 above)
Comment (masteriza) [reply web/src/main/java/ua/com/vertex/dao/CourseDaoImpl.java]: done. don't worry, be happy (=

File: web/src/main/webapp/WEB-INF/views/viewAllCourses.jsp (hunk 11 above)
Comment (masteriza) [reply web/src/main/webapp/WEB-INF/views/viewAllCourses.jsp]: This is not strange. This is the << link to the first page, and this >> to the last page

File: web/src/main/java/ua/com/vertex/utils/Aes.java (hunk 10 above)
Comment (masteriza) [reply web/src/main/java/ua/com/vertex/utils/Aes.java]: Sorry, but this is not my implementation. I do not know the exact purpose of this variable ((

Comment (backumich) [reply web/src/main/java/ua/com/vertex/utils/Aes.java]: then name it "result"

File: web/src/main/webapp/WEB-INF/views/viewAllCourses.jsp (hunk 11 above)
Comment (backumich) [reply web/src/main/webapp/WEB-INF/views/viewAllCourses.jsp]: ok, but can you please take a look on special chars in html? Is it possible to shield it?

File: web/src/main/java/ua/com/vertex/logic/CourseLogicImpl.java [hunk 12]
```diff
@@ -29,11 +28,12 @@ public DataNavigator updateDataNavigator(DataNavigator dataNavigator) {
...
-            if (dataSize / dataNavigator.getCurrentRowPerPage() >= 0) {
-                dataNavigator.setQuantityPages(dataNavigator.getQuantityPages() + 1);
-            }
-            dataNavigator.setLastPage(dataNavigator.getQuantityPages());
+
+            final int totalPages = dataSize / dataNavigator.getRowPerPage();
+            final int quantityPages = dataNavigator.getQuantityPages();
+
+            dataNavigator.setTotalPages(totalPages >= 0 ? quantityPages + 1 : totalPages);
+            dataNavigator.setLastPage(quantityPages);
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/logic/CourseLogicImpl.java]: can we discuss it?

Comment (masteriza) [reply web/src/main/java/ua/com/vertex/logic/CourseLogicImpl.java]: Oh, sure.

File: web/src/main/java/ua/com/vertex/utils/Aes.java (hunk 10 above)
Comment (masteriza) [reply web/src/main/java/ua/com/vertex/utils/Aes.java]: Done

File: web/src/main/java/ua/com/vertex/utils/DataNavigator.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    private int dataSize = 0;
+    private Map<Integer, Integer> countRowPerPage = new TreeMap<Integer, Integer>() {{
+        put(25, 25);
+        put(50, 50);
+        put(100, 100);
+    }};
+
+    private static final Logger LOGGER = LogManager.getLogger(DataNavigator.class);
+
+    public static DataNavigator updateDataNavigator(DataNavigator dataNavigator, int dataSize) {
```
Comment (backumich) [inline web/src/main/java/ua/com/vertex/utils/DataNavigator.java]: 1. static method - solution, but kind of "logic" would be better. 2. Side effect again!!!

Comment (masteriza) [reply web/src/main/java/ua/com/vertex/utils/DataNavigator.java]: Fixed.

**Review by backumich — APPROVED** [review]

**Pull request merged**